### DISCUSSION TITLE
v0.5.0

### DISCUSSION THREAD

**Discussion by sylikc** [sylikc/pyexiftool#40] [General]
Released to [PyPI](https://pypi.org/project/PyExifTool/0.5.0/)

Major refactor / rewrite of ExifTool.  Lots of things changed.  Check COMPATIBILITY.txt for more info.

While it is not directly backwards-compatible with v0.4.x, **most use cases will only require minor changes to your existing code**.

* Python 3.6+ required
* exiftool 12.15+ required

<hr /><em>This discussion was created from the release <a href='https://github.com/sylikc/pyexiftool/releases/tag/v0.5.0'>v0.5.0</a>.</em>

**Comment by csparker247** [comment]
Just letting you know that I finally updated my code to use 0.5.2. The hardest thing was realizing that `ExifTool.execute()` takes str instead of byte strings now. Otherwise no problems!

**Reply by sylikc** [reply]
COOL!  I'm happy to hear.

Well, do you like that change?  I pretty much hated working in bytes... and it caused some weird issues

**Reply by csparker247** [reply]
I certainly think it's more intuitive. I always worked with a list of strings, then just did this before calling execute:

```
args = [a.encode() for a in args]
```

So it doesn't save me much code, but it does save me having to think about it.

**Reply by sylikc** [reply]
Well, thanks for the feedback.

What do you think about the documentation?  Wordy?  Just enough?

It's hard to know how many people actually use the library, what they're doing with it, and what they wished was different :D

I have a few more ideas in the pipeline for new development but they'll all probably be coded in ExifToolAlpha.... otherwise, v0.5 should be pretty stable, API-wise.

**Reply by csparker247** [reply]
I think the docs are great! Not too long or too short. The only thing I would suggest is adding a `docs` badge to the top of the README. I've grown accustomed to having a quick link at the top of a project that takes me straight to the documentation site.

My use cases (two projects) are fairly simple. In one project, I read a few tags from a batch of images. In the other project, I generate and save unique info for each image in a batch of images. So you've got me covered! I'm just happy that there's an active and responsive maintainer when bugs inevitably arise.

**Reply by sylikc** [reply]
> The only thing I would suggest is adding a `docs` badge to the top of the README.

Done.  I updated the README.rst to have a docs badge.

**Reply by csparker247** [reply]
Beautiful!

**Reply by sylikc** [reply]
@csparker247 looks like I'm bringing back ``bytes`` ... apparently the change in v0.5.0 caused unintended side effects, which is ... like can't read bytes as per https://github.com/sylikc/pyexiftool/issues/47 ... and then there was also someone who posted that they can't read files which have mixed encodings...

I'm bringing in the best of both worlds... allow ``bytes`` and ``str`` in params to execute()... and return str by default, or bytes if requested

**Reply by csparker247** [reply]
Thanks for the heads up. I'll keep an eye out for that PR.

**Reply by sylikc** [reply]
Yeah there's a current PR active https://github.com/sylikc/pyexiftool/pull/48 but it's in draft right now so I can think more about whether it may cause unintended consequences itself.

Feel free to review that PR.  The most interesting code you might be interested in is how to extract binary data with the new code https://github.com/sylikc/pyexiftool/pull/48/commits/60d793f78277816fd73051e388dc9f456ca5ad45#diff-3e81514bd7e18cf47803b6172c9276fcc66e03967bd96cdb3d8a2db779ed9db1R30-R60

Another (relatively) big change is that ``ExifToolHelper()`` now allows passing in arbitrary types into ``execute``, which allows something like this which wasn't allowed before:
```python
et.execute("-SomeTag", Path.cwd() / "some file")
```

Which makes it easier... and oh, just reminded me that I can remove some checks in gettags and settags

**Reply by csparker247** [reply]
This seems alright for my purposes. I'm using `ExifTool.execute()`, so supporting `Union[str, bytes]` is sufficient for me. Thanks!
